### ISSUE TITLE
CSVReader May Only Call DelimitedFileReader::Read() Once per Solve.

### ISSUE THREAD

**Issue by crswong888** [idaholab/moose#14785]
## Bug Description
Currently, the `read()` function from `DelimitedFileReader` occurs in `CSVReader::initialize()`. But if a user invokes CSVReader two or more times during a solve it causes data headers in the csv file to be read as normal (double precision) data [on this line](https://github.com/idaholab/moose/blob/7dff289d32368a5f0adb00cd7d723d7bf5e452a7/framework/src/utils/DelimitedFileReader.C#L192). See a more detailed discussion of this bug on a [google groups thread](https://groups.google.com/forum/?hl=en#!topic/moose-users/t3QrFe9bsK4).

## Steps to Reproduce
Set `force_preic = true` or `execute on = 'INITIAL TIMESTEP_END'` with `type = CSVReader` and attempt to read a file with headers on the first line.

## Impact
It is impossible to use `force_preic = true` in the current implementation of CSVReader if the file to be read contains headers.

## Proposed Fix
By calling `DelimitedFileReader::read()` only once, and declaring the vector variables once as well, the issue no longer occurs. This can be done by carrying out these processes in the constructor rather than in `initialize()`. For example:

```
CSVReader::CSVReader(const InputParameters & params)
  : GeneralVectorPostprocessor(params), _csv_reader(getParam<FileName>("csv_file"), &_communicator)
{
  _csv_reader.read();
  // declare Vectors from data headers
  for (auto & name : _csv_reader.getNames())
    if (_column_data.find(name) == _column_data.end())
      _column_data[name] = &declareVector(name);
}

void
CSVReader::initialize()
{
  // declareVector() previously occured here...
  // when initialize() was invoked twice it cause
  // DelimitedFileReader to read the header line
  // as normal data on the second invocation
}
```

On a side note, CSVReader currently calls the `DelimitedFileReader::setHeaderFlag()` function, but this code is deprecated. The new `DelimitedFileReader::header()` function automatically determines if there is a header or not and therefore the `params.addParam<bool>("header", " ")` input is no longer necessary.

@permcody @aeslaughter @YaqiWang

**Comment by permcody** [comment]
Conceptually, I would argue it doesn't make sense to call a file reader that's reading an invariant file more than once. So rather than muck up the FileReader to handle this case, work should be performed to make sure that it just doesn't get called twice in any case.

You might note that I said I really didn't like the "force_preic" parameter when merging that in. It's definitely a hack and these kinds of issues are precisely why I don't like those kinds of hacks.

That being said, the cat is out of the bag and this needs to be resolved. I don't really have the resources to fix this right now so if anyone else wants to take a crack at, go nuts!

**Comment by crswong888** [comment]
@permcody, I hear you. I can fix this code in the way you suggested. I agree that the same exact csv data need not be read more than once.

**Comment by aeslaughter** [comment]
This was closed by #14892. I also am apparently referencing this issue in another PR by mistake.
